aten/src/ATen/mps/MPSDevice.mm

```diff
@@ -3,26 +3,72 @@
 #include <c10/util/CallOnce.h>
 
 #include <ATen/mps/MPSDevice.h>
+#include <ATen/mps/IndexKernels.h>
 
 namespace at {
 namespace mps {
 
 static std::unique_ptr<MPSDevice> mps_device;
 static c10::once_flag mpsdev_init;
 
+static inline MTLLanguageVersion getMetalLanguageVersion(const id<MTLDevice>& device) {
+  // MPS Advanced Indexing needs at least Metal 2.0 (support for Argument Buffers and function constants)
+  MTLLanguageVersion languageVersion;
+
+#if defined(__MAC_13_0) && __MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_13_0
+  languageVersion = MTLLanguageVersion3_0;
+#elif defined(__MAC_12_0) && __MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_12_0
+  languageVersion = MTLLanguageVersion2_4;
+#elif
+  #error "Metal is not available on the current platform."
+#endif
+
+  TORCH_CHECK([device supportsFamily:MTLGPUFamilyMac2], "Missing Metal support for MTLGPUFamilyMac2");
```

Comment:
This call is made every time `index.Tensor_out` graph is generated, although it is a function of a device. Please move this check to `MTLDevice` constructor

Reply:
No, this wouldn't be called for every `index.Tensor_out` call. The `_mtl_indexing_library` is cached on the `MPSDevice` the very first time an `index.Tensor_out` is called. All the subsequent calls made afterwards will use directly the cached version from the `MPSDevice` (they won't go again through this code).

```
id<MTLFunction> MPSDevice::metalIndexingFunction(const std::string& kernel, MTLFunctionConstantValues* constantValues) {
  assert(_mtl_device);
  NSError* error = nil;
  if (!_mtl_indexing_library) {  // Create the library only if it was not already previously created
    MTLCompileOptions *options = [MTLCompileOptions new];
    [options setLanguageVersion: getMetalLanguageVersion(_mtl_device)];
    [options setFastMathEnabled: YES];
    _mtl_indexing_library = [_mtl_device newLibraryWithSource: [NSString stringWithCString: mps::indexing_metal_shaders encoding:NSASCIIStringEncoding]
                                                      options: options
                                                        error: &error];
    TORCH_CHECK(_mtl_indexing_library, "Failed to create indexing library, error: ", [[error description] UTF8String]);
  }
  ...
```